compiler/src/abi/json_builder.rs

```diff
@@ -0,0 +1,240 @@
+            _ => None,
+        })
+        .collect();
+
+    module
+        .body
+        .iter()
+        .filter_map(|stmt| match stmt.node {
+            vyp::ModuleStmt::ContractDef { .. } => Some(contract_def(&type_defs, &stmt.node)),
+            _ => None,
+        })
+        .collect::<Result<Vec<Contract>, CompileError>>()
+}
+
+/// Builds a contract ABI.
+pub fn contract_def<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    stmt: &'a vyp::ModuleStmt<'a>,
+) -> Result<Contract, CompileError> {
+    if let vyp::ModuleStmt::ContractDef { name: _, body } = stmt {
+        return Ok(Contract {
+            functions: body
+                .iter()
+                .map(|stmt| func_def(type_defs, &stmt.node))
+                .collect::<Result<Vec<Option<Function>>, CompileError>>()?
+                .into_iter()
+                .filter_map(|function| function)
+                .collect(),
+        });
+    }
+
+    Err(CompileError::static_str(
+        "Contract definition translation requires ContractDef parameter.",
+    ))
+}
+
+/// Builds an ABI function element, returning None if the function is private.
+pub fn func_def<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    stmt: &'a vyp::ContractStmt<'a>,
+) -> Result<Option<Function>, CompileError> {
+    if let vyp::ContractStmt::FuncDef {
+        qual,
+        name,
+        args,
+        return_type,
+        body: _,
+    } = stmt
+    {
+        if qual.is_none() {
+            return Ok(None); // Private method.
+        }
+
+        let outputs = if let Some(return_type) = return_type {
+            vec![Output {
+                name: String::from("return_value"),
+                typ: type_desc(type_defs, &return_type.node)?,
+            }]
+        } else {
+            Vec::new()
+        };
+
+        return Ok(Some(Function {
+            name: String::from(name.node),
+            typ: FunctionType::Function,
+            inputs: args
+                .iter()
+                .map(|arg| func_def_arg(type_defs, &arg.node))
+                .collect::<Result<Vec<Input>, CompileError>>()?,
+            outputs,
+        }));
+    }
+
+    Err(CompileError::static_str(
+        "Function definition translation requires FuncDef parameter.",
+    ))
+}
+
+/// Builds a function input.
+pub fn func_def_arg<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    arg: &'a vyp::FuncDefArg<'a>,
+) -> Result<Input, CompileError> {
+    Ok(Input {
+        name: String::from(arg.name.node),
+        typ: type_desc(&type_defs, &arg.typ.node)?,
+    })
+}
+
+/// Maps the type description to an ABI-friendly type and handles custom types.
+pub fn type_desc<'a>(
+    type_defs: &'a TypeDefs<'a>,
+    typ: &'a vyp::TypeDesc<'a>,
+) -> Result<VariableType, CompileError> {
+    if let vyp::TypeDesc::Base { base } = typ {
+        if let Some(custom_type) = type_defs.get(base) {
+            return type_desc(&HashMap::new(), custom_type);
```

Comment:
why pass a new `type_defs` here (via `HashMap::new()`) and not supply the function's `type_defs`... seems like we are losing the capacity for embedded custom types this way...

Reply:
Yeah, not sure what the thinking was here.